### PR TITLE
Fix type narrowing for type(x) expressions with type[T] parameters and reverse comparisons

### PR DESCRIPTION
**Pull request by hsusul**
## Summary

Fixes a type narrowing issue where binary comparison type guards involving `type(x)` expressions failed to narrow types when:
1. `type(x)` appeared on the right-hand side of the comparison (e.g. `if cls is type(x):` or `if int is type(x):`).
2. The comparison operand was a variable or parameter of type `type[T]` (e.g. `cls: type[int]`).

## Root Cause
1. Pyright's binary type guard logic for `type(x)` previously checked only `testExpression.d.leftExpr.nodeType === ParseNodeType.Call`. When `type(x)` was on the RHS, Pyright skipped generating a type guard callback.
2. When comparing `type(x)` against a variable `cls` of type `type[T]`, `getTypeOfExpression(cls)` returns a `ClassType` instance of `type`. Pyright checked `isInstantiableClass(expandedSubtype)`, which returned `false` for `type[T]` instances, preventing `classTypes` from being populated.
3. Positive narrowing in `narrowTypeForTypeIs` did not eliminate disjoint subtypes when comparing against class types derived from `type[T]`.

## Changes
- Supported `type(x)` on either `leftExpr` or `rightExpr` of `is`, `is not`, `==`, and `!=` binary comparison operators.
- Unwrapped `type[T]` instances (`ClassType` built-in `type` with `typeArgs`) to instantiable class types (`convertToInstantiable(typeArgs[0], false)`).
- Updated positive narrowing logic in `narrowTypeForTypeIs` to check subclass overlap and eliminate disjoint subtypes.
- Excluded `type[object]` operands to preserve correct behavior when comparing against generic `object` instances.
- Added regression tests in `typeIs6.py` covering all `type(x)` comparison forms, `type[T]` parameters, `@final` classes, and union types.

### CODE REVIEW

**Comment by StellaHuang95** [comment]
🔒 Automated review in progress — @stellahuang95 is auto-reviewing this PR.

**Comment by StellaHuang95** [comment]
Negative `type(...) is not ...` narrowing can eliminate valid supertypes, producing unsound results. The PR needs this corrected and covered by a regression test.

**Comment by StellaHuang95** [comment]
The negative narrowing change can incorrectly eliminate a valid superclass subtype when the compared class is a final subclass, producing unsound results.

**Comment by StellaHuang95** [comment]
The negative narrowing change can incorrectly eliminate a non-final base class when the compared class is a final subclass, producing an unsound narrowed type.

**Comment by hsusul** [comment]
Thank you for the review and sharp catch @StellaHuang95!

I have updated the negative narrowing logic in `narrowTypeForTypeIs` to check `ClassType.isFinal(instantiableSubtype)` rather than `ClassType.isFinal(classType)`. This ensures that a non-final base class (such as `Base`) is preserved when comparing `type(x) is not FinalSub` against a `@final` subclass (`FinalSub`), preserving sound type narrowing when `x` is an instance of `Base`.

I have pushed the fix and added a regression test (`func8`) in `typeIs6.py` covering this exact pattern:
```python
class Base: pass

@final
class FinalSub(Base): pass

def func8(x: Base):
    if type(x) is not FinalSub:
        reveal_type(x, expected_text="Base")
    else:
        reveal_type(x, expected_text="FinalSub")
```

**Comment by StellaHuang95** [comment]
GitHub cannot anchor PR review comments to unchanged lines in the diff. Falling back to a general PR comment for [packages/pyright-internal/src/analyzer/typeGuards.ts:L2626](https://github.com/microsoft/pyright/blob/0d980d43ea441d7dfd5eba19e798f601fb858ea6/packages/pyright-internal/src/analyzer/typeGuards.ts#L2626).

> **Info** · Optional note

`classType` is always created with `includeSubclasses` set to `false` at its sole call site, so the first disjunct always short-circuits and `isSubclass` cannot affect narrowing. Please remove this inert condition and computation, or add a reachable subclass-inclusive case with a comment explaining the intended future behavior.

File: packages/pyright-internal/src/analyzer/typeGuards.ts [hunk 1]
```diff
@@ -209,45 +210,76 @@ export function getTypeNarrowingCallback(
...
+                        if (isClassType && classTypes.length > 0) {
+                            return (type: Type) => {
+                                return {
+                                    type: narrowTypeForTypeIs(evaluator, type, classTypes, adjIsPositiveTest),
+                                    isIncomplete: !!otherResult.isIncomplete,
+                                };
+                            };
                         }
                     }
                 }
```
Comment (StellaHuang95) [inline packages/pyright-internal/src/analyzer/typeGuards.ts:292]: > **Warning** · Non-blocking recommendation

The blanket `object` exclusion also changes direct `type(x) is object` checks from narrowing the positive branch to `Never` to performing no narrowing. It also disables the entire callback for `type[int] | type[object]`, losing the useful `int` narrowing. Please scope the exclusion to the unwrapped `type[object]` case, or add tests documenting that this broader precision loss is intentional.

File: packages/pyright-internal/src/tests/samples/typeIs6.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+class Base: pass
+
+@final
+class FinalSub(Base): pass
+
+def func8(x: Base):
+    if type(x) is not FinalSub:
+        reveal_type(x, expected_text="Base")
+    else:
+        reveal_type(x, expected_text="FinalSub")
```
Comment (StellaHuang95) [inline packages/pyright-internal/src/tests/samples/typeIs6.py:60]: > **Warning** · Non-blocking recommendation

Please add product-path regression coverage under `packages/pylance-internal/src/tests` for this shared narrowing behavior. Also add the negative case `x: int | str; if type(x) is bool: ... else: reveal_type(x, expected_text="int | str")` to guard against incorrectly eliminating a valid supertype when the compared class is a final subclass.

**Review by StellaHuang95 — DISMISSED** [review]
Approved via Review Center.

**Comment by rchiodo** [comment]
I took a closer look at the `mypy_primer` differences. I don't think all of them are expected, and I would not merge this as-is.

The main issue is that the new logic converts the argument of `type[T]` with `includeSubclasses=false`. A value annotated `type[T]` can be a class object for a subclass of `T` unless `T` is final, so treating it as exactly `T` is unsound.

* **Spark:** This looks like a real regression. In `MultiIndex.symmetric_difference`, `other` is annotated as `Index`, and the code checks `if type(self) is not type(other): raise`. The continuation is reachable when `other` is a `MultiIndex`, but this PR narrows `self` to `Never`, producing the two new errors.
* **koda-validate:** The three new errors are false positives. The reverse comparison is now recognized, but narrowing loses the correlation between the constrained `ExactMatchT` and `val` after `type(self.match) == type(val)`.
* **yarl:** The removed error is an expected and desirable improvement. `if type(val) is cls: return val` should narrow `val` sufficiently for the `URL` return type.
* **SymPy:** The large cascade appears to be primer/inference noise rather than a reproducible semantic difference from this PR. I would rerun primer rather than accept or update expectations based on that block.

I recommend preserving subclass inclusion for operands typed `type[T]` (using exact-class treatment only for direct class expressions or final classes), and adding a Spark-shaped regression test that verifies the continuation does not become `Never`. It would also be useful to distinguish `Base` from `cls: type[Base]` in the tests.

**Comment by hsusul** [comment]
Thank you for the detailed `mypy_primer` analysis and architectural guidance!

I have updated the binary type guard unwrapping logic in `getTypeNarrowingCallback`:
1. **Operands typed `type[T]`** (e.g. `cls: type[Index]` or `type(other)` where `other: Index`): We preserve subclass inclusion (`includeSubclasses = true`) when converting `type[T]`. This ensures that for `type(self) is not type(other)` (where `self: MultiIndex` and `other: Index`), `type(other)` can represent `MultiIndex` or any subclass of `Index`, so `self` is not unsoundly narrowed to `Never` on continuation (resolving the Spark `MultiIndex.symmetric_difference` regression).
2. **Direct Class Expressions** (e.g. `if type(x) is Base:` or `if int is type(x):`): We retain exact-class treatment (`includeSubclasses = false`).

### Added Tests in `typeIs6.py`
- **Spark Regression Test** (`test_spark_regression`): Verifies `self` remains `MultiIndex` (does not become `Never`) after `if type(self) is not type(other): pass`.
- **Direct Class vs `type[Base]` Test** (`test_direct_class_vs_type_param`): Verifies `Base` handling for direct class expressions vs `cls: type[Base]` parameters.

The update has been pushed in commit `94afc46b3`.

File: packages/pyright-internal/src/analyzer/typeGuards.ts [hunk 3]
```diff
@@ -209,45 +210,83 @@ export function getTypeNarrowingCallback(
...
```
Comment (StellaHuang95) [inline packages/pyright-internal/src/analyzer/typeGuards.ts:292]: > **Warning** · Non-blocking recommendation

The `object` exclusion now also rejects the pre-existing direct `type(x) is object` form. Previously it produced an impossible positive branch (`Never`); now `isClassType` becomes false and no narrowing callback is returned. It also discards useful alternatives in unions such as `type[int] | type[object]`. Please exclude only the parameterized `type[object]` case, or skip that subtype rather than disabling the entire callback.

File: packages/pyright-internal/src/analyzer/typeGuards.ts [hunk 4]
```diff
@@ -209,45 +210,83 @@ export function getTypeNarrowingCallback(
...
+                    const callType = evaluator.getTypeOfExpression(
+                        typeCallNode.d.leftExpr,
+                        EvalFlags.CallBaseDefaults
+                    ).type;
+
+                    if (isInstantiableClass(callType) && ClassType.isBuiltIn(callType, 'type')) {
+                        const otherResult = evaluator.getTypeOfExpression(otherExpr);
+                        const classTypes: ClassType[] = [];
+                        let isClassType = true;
+
```
Comment (StellaHuang95) [inline packages/pyright-internal/src/analyzer/typeGuards.ts:243]: > **Info** · Optional note

When both operands are calls, the left call is selected before checking whether its argument matches the reference. As a result, querying narrowing for `x` in `type(y) is type(x)` stops after `type(y)` fails the match and never considers the RHS `type(x)`. Please fall through to the RHS call when the selected call does not match the reference.

**Review by StellaHuang95 — APPROVED** [review]
Approved via Review Center.

**Review by rchiodo — APPROVED** [review]
Approved via Review Center.

**Comment by rchiodo** [comment]
🔒 Automated review in progress — @rchiodo is auto-reviewing this PR.

File: packages/pyright-internal/src/analyzer/typeGuards.ts [hunk 5]
```diff
@@ -209,45 +210,83 @@ export function getTypeNarrowingCallback(
...
-                                    } else {
-                                        isClassType = false;
+                typeCallNode = testExpression.d.leftExpr;
+                otherExpr = testExpression.d.rightExpr;
+            } else if (testExpression.d.rightExpr.nodeType === ParseNodeType.Call) {
+                typeCallNode = testExpression.d.rightExpr;
+                otherExpr = testExpression.d.leftExpr;
+            }
+
+            if (
```
Comment (rchiodo) [inline packages/pyright-internal/src/analyzer/typeGuards.ts:226]: > **Warning** · Non-blocking recommendation

The left-first `if/else if` selection skips a valid RHS `type(reference)` when the LHS is also a call, such as `factory() is type(x)` or `type(y) is type(x)`. Check both operands independently and add a reverse call/call regression test.

File: packages/pyright-internal/src/analyzer/typeGuards.ts [hunk 6]
```diff
@@ -209,45 +210,83 @@ export function getTypeNarrowingCallback(
...
+                                    ) {
+                                        isTypeParam = true;
+                                        const extracted = convertToInstantiable(
+                                            expandedSubtype.priv.typeArgs[0],
+                                            /* includeSubclasses */ true
+                                        );
+                                        if (isInstantiableClass(extracted)) {
+                                            instantiable = extracted;
+                                        }
+                                    } else if (isInstantiableClass(expandedSubtype)) {
```
Comment (rchiodo) [inline packages/pyright-internal/src/analyzer/typeGuards.ts:265]: > **Warning** · Non-blocking recommendation

Direct classes are cloned with `includeSubclasses=false`, while extracted `type[T]` classes retain subclass inclusion. Add focused tests combining both operand representations to demonstrate that their differing downstream elimination behavior is sound.

**Review by rchiodo — APPROVED** [review]
Approved via Review Center.

**Review by heejaechang — APPROVED** [review]
Approved via Review Center.

File: packages/pyright-internal/src/analyzer/typeGuards.ts [hunk 7]
```diff
@@ -209,45 +210,83 @@ export function getTypeNarrowingCallback(
...
-                                        classTypes.push(expandedSubtype);
-                                    } else {
-                                        isClassType = false;
+                typeCallNode = testExpression.d.leftExpr;
+                otherExpr = testExpression.d.rightExpr;
+            } else if (testExpression.d.rightExpr.nodeType === ParseNodeType.Call) {
+                typeCallNode = testExpression.d.rightExpr;
+                otherExpr = testExpression.d.leftExpr;
+            }
+
```
Comment (rchiodo) [inline packages/pyright-internal/src/analyzer/typeGuards.ts:225]: > **Issue** · Please address or respond

When the left operand is any call, it is selected before verifying that it is `type(reference)`. As a result, `get_class() is type(x)` and `type(y) is type(x)` cannot create a guard for `x`, despite `type(x)` being on the RHS. If the selected call does not match the reference, inspect the other operand as well.

File: packages/pyright-internal/src/analyzer/typeGuards.ts [hunk 8]
```diff
@@ -2538,17 +2577,20 @@ function narrowTypeForTypeIs(evaluator: TypeEvaluator, type: Type, classTypes: C
...
+                            if (ClassType.isSameGenericClass(instantiableSubtype, classType)) {
                                 return addConditionToType(subtype, getTypeCondition(classType));
                             }
 
                             return addConditionToType(ClassType.cloneAsInstance(classType), subtype.props?.condition);
                         }
 
-                        if (!classType.priv.includeSubclasses) {
+                        if (!classType.priv.includeSubclasses || !isSubclass) {
                             return undefined;
```
Comment (rchiodo) [inline packages/pyright-internal/src/analyzer/typeGuards.ts:2594]: > **Issue** · Please address or respond

This removes an `A` subtype for a positive comparison against `type[B]` whenever `A` is not a subclass of `B`, but that does not establish disjointness. A non-final `C(A, B)` can satisfy the comparison when `cls` is `C`, so this narrowing makes a valid runtime type impossible. Preserve such bases unless overlap is impossible, and add a multiple-inheritance regression test.

File: packages/pyright-internal/src/tests/samples/typeIs6.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    if type(self_val) is not type(other):
+        pass
+    reveal_type(self_val, expected_text="MultiIndex")
+
+def test_direct_class_vs_type_param(x: Base, cls: type[Base]):
+    if type(x) is not Base:
+        reveal_type(x, expected_text="Base")
+    
+    if type(x) is not cls:
+        reveal_type(x, expected_text="Base")
```
Comment (rchiodo) [inline packages/pyright-internal/src/tests/samples/typeIs6.py:76]: > **Warning** · Non-blocking recommendation

The regression coverage does not exercise all newly supported operand/operator paths: reversed `is not` and `==`, both `!=` orientations, two competing `type(...)` calls, literal `object`, and multiple-inheritance overlap are missing. Add these cases so the new branches and edge conditions remain protected.

**Comment by rchiodo** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

**Comment by heejaechang** [comment]
### 🔴 Pyright CLI QA — `94afc46b` — verdict: red

Automated CLI checks compared this change with its base revision.
13 focused check(s) completed.

#### Needs review: Diagnostics changed for this example

The base and PR report different diagnostics for the same example. Is this change intended?

<details><summary>Reproduction and observed results</summary>

Create these files in an empty directory:

**`main.py`**

```python
from typing import reveal_type

def check(value: bool, kind: type[int]) -> int:
    if type(value) is kind:
        reveal_type(value)
        return "not an int"
    return 0

check(True, bool)
```

**`pyrightconfig.json`**

```json
{"include": ["main.py"], "pythonVersion": "3.13", "typeCheckingMode": "standard", "reportUnreachable": "warning"}
```

From that directory, run each built Pyright revision:

```text
node <pyright-cli-entrypoint> --outputjson -p .
```

**Base `eefd22541da12c4d5e7703c4ab9d1561131cacd9`**

```json
{
  "exit_code": 1,
  "diagnostics": [
    {
      "file": "main.py",
      "severity": "error",
      "rule": "reportReturnType",
      "message": "Type \"Literal['not an int']\" is not assignable to return type \"int\"\n  \"Literal['not an int']\" is not assignable to \"int\"",
      "range": {
        "start": [
          5,
          15
        ],
        "end": [
          5,
          27
        ]
      }
    },
    {
      "file": "main.py",
      "severity": "information",
      "rule": null,
      "message": "Type of \"value\" is \"bool\"",
      "range": {
        "start": [
          4,
          20
        ],
        "end": [
          4,
          25
        ]
      }
    }
  ]
}
```

**PR `94afc46b3c36cd7eb40c6723e15b9e90a6168cd6`**

```json
{
  "exit_code": 0,
  "diagnostics": [
    {
      "file": "main.py",
      "severity": "warning",
      "rule": "reportUnreachable",
      "message": "Type analysis indicates code is unreachable",
      "range": {
        "start": [
          4,
          8
        ],
        "end": [
          4,
          19
        ]
      }
    }
  ]
}
```

Diagnostic ranges use zero-based line and character positions.

</details>

#### Needs review: `reportReturnType` newly reported

The PR reports `reportReturnType` for this example; the base does not. Is this change intended?

<details><summary>Reproduction and observed results</summary>

Create these files in an empty directory:

**`main.py`**

```python
from typing import reveal_type

def require_exact[T](value: object, kind: type[T]) -> T:
    if type(value) is kind:
        reveal_type(value)
        return value
    raise TypeError

reveal_type(require_exact(1, int))
```

**`pyrightconfig.json`**

```json
{"include": ["main.py"], "pythonVersion": "3.13", "typeCheckingMode": "standard", "reportUnreachable": "warning"}
```

From that directory, run each built Pyright revision:

```text
node <pyright-cli-entrypoint> --outputjson -p .
```

**Base `eefd22541da12c4d5e7703c4ab9d1561131cacd9`**

```json
{
  "exit_code": 0,
  "diagnostics": [
    {
      "file": "main.py",
      "severity": "information",
      "rule": null,
      "message": "Type of \"require_exact(1, int)\" is \"int\"",
      "range": {
        "start": [
          8,
          12
        ],
        "end": [
          8,
          33
        ]
      }
    },
    {
      "file": "main.py",
      "severity": "information",
      "rule": null,
      "message": "Type of \"value\" is \"object*\"",
      "range": {
        "start": [
          4,
          20
        ],
        "end": [
          4,
          25
        ]
      }
    }
  ]
}
```

**PR `94afc46b3c36cd7eb40c6723e15b9e90a6168cd6`**

```json
{
  "exit_code": 1,
  "diagnostics": [
    {
      "file": "main.py",
      "severity": "error",
      "rule": "reportReturnType",
      "message": "Type \"object\" is not assignable to return type \"T@require_exact\"\n  Type \"object\" is not assignable to type \"T@require_exact\"",
      "range": {
        "start": [
          5,
          15
        ],
        "end": [
          5,
          20
        ]
      }
    },
    {
      "file": "main.py",
      "severity": "information",
      "rule": null,
      "message": "Type of \"require_exact(1, int)\" is \"int\"",
      "range": {
        "start": [
          8,
          12
        ],
        "end": [
          8,
          33
        ]
      }
    },
    {
      "file": "main.py",
      "severity": "information",
      "rule": null,
      "message": "Type of \"value\" is \"object\"",
      "range": {
        "start": [
          4,
          20
        ],
        "end": [
          4,
          25
        ]
      }
    }
  ]
}
```

Diagnostic ranges use zero-based line and character positions.

</details>
